### PR TITLE
Restore support for the `include/pattern` syntax.

### LINKED ISSUE #22933: Include and exclude patterns in visualizations throw an error
**Elasticsearch version**: 5.2.0

As of 5.2.0, using an object syntax for `exclude` in a terms aggregation will result in a 400 `illegal_argument_exception` error.  The exact same query works in 5.1.2.

Example _msearch request body (taken directly from Kibana):
```
{"index":["logstash-0"],"ignore_unavailable":true,"preference":1486048204582}
{"size":0,"query":{"bool":{"must":[{"query_string":{"analyze_wildcard":true,"query":"*"}},{"range":{"@timestamp":{"gte":1486047452495,"lte":1486048352495,"format":"epoch_millis"}}}],"must_not":[]}},"_source":{"excludes":[]},"aggs":{"2":{"terms":{"field":"referer","exclude":{"pattern":"apollo"},"size":5,"order":{"_count":"desc"}}}}}
```

Response body:
```
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"[terms] exclude doesn't support values of type: START_OBJECT"}],"type":"illegal_argument_exception","reason":"[terms] exclude doesn't support values of type: START_OBJECT"},"status":400}
```

This was originally filed against Kibana: https://github.com/elastic/kibana/issues/10153

@javanna believes it may have been caused by https://github.com/elastic/elasticsearch/pull/22048

### PR DESCRIPTION
**Pull request by jpountz**
Closes #22933

### CODE REVIEW

**Comment by jpountz** [comment]
@markharwood Would you mind reviewing this PR?

**Comment by markharwood** [comment]
Just tried this regex which works OK: 

	GET test/_search
	{
		"size":0,
		"aggs":{
			"terms":{
				"terms":{
					"field":"foo.keyword",
					"include":"b.*",                
					"exclude":"a.*"				
				}
			}
		}
	}

.. but this equivalent syntax fails to parse: 

	GET test/_search
	{
		"size":0,
		"aggs":{
			"terms":{
				"terms":{
					"field":"foo.keyword",
					"include":{
						"pattern":"ba.*"                    
					},
					"exclude":{ 
						"pattern":"a.*"
					}
				}
			}
		}
	}

**Comment by jpountz** [comment]
Thanks @markharwood I should have been more surprised that tests passed on the first time that I ran them, I probably ran gradle from the wrong directory, tests were completely broken! :) I just pushed a new commit that should fix it.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java [hunk 1]
```diff
@@ -120,6 +121,9 @@ public static IncludeExclude parseInclude(XContentParser parser, QueryParseConte
                             "Unknown parameter in Include/Exclude clause: " + currentFieldName);
                 }
             }
+            if (pattern != null) {
+                return new IncludeExclude(pattern, null);
```
Comment (markharwood) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java]: Maybe throw an exception here if partition criteria supplied?
Partitioning not supported when using regex-based filtering.

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/support/IncludeExclude.java]: good point

**Comment by jpountz** [comment]
@markharwood Can you give it another look?

**Comment by markharwood** [comment]
LGTM

**Pull request merged**